lib/rebaser-server/src/server/core_loop.rs

```diff
@@ -0,0 +1,190 @@
+use dal::{
+    DalContext, DalContextBuilder, JobQueueProcessor, ServicesContext, Tenancy, TransactionsError,
+    Visibility, WorkspacePk,
+};
+use futures::FutureExt;
+use futures::StreamExt;
+use rebaser_core::{
+    RebaserMessagingConfig, ReplyRebaseMessage, RequestRebaseMessage, SubjectGenerator,
+};
+use si_crypto::SymmetricCryptoService;
+use si_data_nats::jetstream::{AckKind, JetstreamError, Stream, REPLY_SUBJECT_HEADER_NAME};
+use si_data_nats::subject::ToSubject;
+use si_data_nats::NatsClient;
+use si_data_pg::PgPool;
+use std::sync::Arc;
+use std::time::Instant;
+use stream_cancel::StreamExt as CancelStreamExt;
+use telemetry::prelude::*;
+use thiserror::Error;
+use tokio::sync::watch;
+
+use crate::server::rebase::perform_rebase;
+
+#[remain::sorted]
+#[derive(Debug, Error)]
+pub enum CoreLoopSetupError {
+    #[error("jetstream error: {0}")]
+    Jetstream(#[from] JetstreamError),
+    #[error("serde json erorr: {0}")]
+    SerdeJson(#[from] serde_json::Error),
+    #[error("transactions error: {0}")]
+    Transactions(#[from] Box<TransactionsError>),
+}
+
+impl From<TransactionsError> for CoreLoopSetupError {
+    fn from(e: TransactionsError) -> Self {
+        Self::Transactions(Box::new(e))
+    }
+}
+
+type CoreLoopSetupResult<T> = Result<T, CoreLoopSetupError>;
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) async fn setup_and_run_core_loop(
+    pg_pool: PgPool,
+    nats: NatsClient,
+    veritech: veritech_client::Client,
+    job_processor: Box<dyn JobQueueProcessor + Send + Sync>,
+    symmetric_crypto_service: SymmetricCryptoService,
+    encryption_key: Arc<veritech_client::CycloneEncryptionKey>,
+    shutdown_watch_rx: watch::Receiver<()>,
+    messaging_config: RebaserMessagingConfig,
+    content_store_pg_pool: PgPool,
+) -> CoreLoopSetupResult<()> {
+    let services_context = ServicesContext::new(
+        pg_pool,
+        nats.clone(),
+        job_processor,
+        veritech.clone(),
+        encryption_key,
+        None,
+        None,
+        symmetric_crypto_service,
+        messaging_config.clone(),
+        content_store_pg_pool,
+    );
+
+    // Setup the subjects.
+    let subject_all = SubjectGenerator::all(messaging_config.subject_prefix());
+    let subject_root = SubjectGenerator::root(messaging_config.subject_prefix());
+    info!(%subject_all, %subject_root, "created services context and prepared subjects");
+
+    // Setup the stream and the consumer.
+    let jetstream_ctx = nats.clone().to_jetstream_ctx();
+    let rebaser_jetstream_stream = jetstream_ctx
+        .get_or_create_stream(&subject_root, vec![subject_all])
+        .await?;
```

Comment:
Like before, the rebaser server(s) will try to find or create the stream they consume on. In the future, we should consider other services trying to do this too (like pinga and sdf) for rolling deployments.